### ISSUE TITLE
Linux: error: unknown token in expression

### ISSUE THREAD

**Issue by Nassiel** [dotnet/coreclr#77]
SO: Centos 7 - kernel 3.10.0-123.20
Cmake 3.1.1
Clang 3.4.2
gcc 4.8.2
g++ 4.8.2

This is the error that throws me while compiling the coreclr:
/tmp/jithelpers_fastwritebarriers-ca1b4c.s:464:21: error: unknown token in expression
        movabs rax, 0F0F0F0F0F0F0F0F0h

Could be related with:
fatal error: 'stropts.h' file not found

Linux doesn't support STREAMS almost never. POSIX recommends this check
#include <unistd.h>
#if defined _XOPEN_STREAMS && _XOPEN_STREAMS == -1
/* XSR option is not available, headers, data types etc. may not be available. */
#endif

**Comment by shahid-pk** [comment]
you need to have clang and llvm 3.5

**Comment by ArkQA** [comment]
3.5 works for me. (Out of curiosity, are you using package repos? I thought gcc/g++ was 4.9+ already)

**Comment by Nassiel** [comment]
Yes all package repos, but I'll build from the latest instead of repos. Despite of that, I think that it's an error force to use the latest version of each component.

**Comment by cnblogs-dudu** [comment]
I build CoreCLR on Linux CentOS 6.5 and experience following errors:
```
Executing make
[  0%] Building CXX object src/pal/src/CMakeFiles/CoreClrPal.dir/exception/seh.cpp.o
In file included from /data/git/coreclr/src/pal/src/exception/seh.cpp:22:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/typeinfo:34:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/exception:148:
/usr/bin/../lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/exception_ptr.h:143:13: error: unknown type
      name 'type_info'
      const type_info*
            ^
1 error generated.
make[2]: *** [src/pal/src/CMakeFiles/CoreClrPal.dir/exception/seh.cpp.o] Error 1
make[1]: *** [src/pal/src/CMakeFiles/CoreClrPal.dir/all] Error 2
make: *** [all] Error 2
Failed to build coreclr components.
```

**Comment by shahid-pk** [comment]
@cnblogs-dudu 
the only supported compiler currently is clang 3.5 , and i think somehow your build was compiling it using gcc.

**Comment by cnblogs-dudu** [comment]
@shahid-pk 
thanks. `yum install clang` on CentOS, the latest version of clang is 3.4.2. Maybe this is reason. I will try Ubuntu.

**Comment by cnblogs-dudu** [comment]
Manually install clang 3.5 on Ubuntu and build successfully!
* `apt-get install cmake`
* `wget http://llvm.org/releases/3.5.0/clang+llvm-3.5.0-x86_64-linux-gnu-ubuntu-14.04.tar.xz`
* `tar xvf clang+llvm-3.5.0-x86_64-linux-gnu-ubuntu-14.04.tar.xz`
* `cd clang+llvm-3.5.0-x86_64-linux-gnu`
* `cp -R * /usr`
* `sh build.sh`

**Comment by Nassiel** [comment]
I'm trying in different sites, one of them is still compiling llvm with clang for 5 hours, the other is compiled and the default compiler established to clang cc and c++ in order to make them default to CMake.

Despite of that, the first try gives that:
error: unknown type name 'char16_t' typedef chat16_t wchar_t;

**Comment by Nassiel** [comment]
Ok, now clang 3.5.1, cmake 3.1.1, I tried in 4 different centos and each one receives a different error but the most advanced is this, so I imagine that is question of configuration:

Scannin dependecies of target mdhotdata_full
src/init//pal.cpp:63:10: fatal error: 'stropts.h' file not found
Linking CXX static library libmdhotdata_full.a
Build  target mdhotdata_full
make: *** [all] Error 2

**Comment by sergiy-k** [comment]
@Nassiel The issue with 'stropts.h' has been fixed by #82.

**Comment by Nassiel** [comment]
Finally works perfect, Thanks to all!
